File: src/trading_ai/ui/models/interactive_portfolio.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class PortfolioPosition(BaseModel):
    position_id: str
    account_id: str = "paper-account"
    symbol: str
    instrument_type: Literal["EQUITY", "OPTION"] = "OPTION"
    quantity: float
    average_price: float = 0
    mark_price: float = 0
    multiplier: float = 100
    option_expiry: str | None = None
    option_strike: float | None = None
    option_type: Literal["CALL", "PUT"] | None = None
    delta: float = 0
    gamma: float = 0
    theta: float = 0
    vega: float = 0
    implied_volatility: float = 0
    underlying_price: float = 0
    market_value: float = 0
    unrealized_pnl: float = 0


class PortfolioSummary(BaseModel):
    generated_at: datetime
    account_id: str
    total_market_value: float
    total_unrealized_pnl: float
    gross_exposure: float
    net_exposure: float
    net_delta: float
    net_gamma: float
    net_theta: float
    net_vega: float
    long_positions: int
    short_positions: int
    symbols: int
    positions: list[PortfolioPosition]


class ScenarioPoint(BaseModel):
    underlying_shock_pct: float
    volatility_shock_points: float
    days_forward: int
    estimated_pnl: float
    estimated_market_value: float
    delta_pnl: float
    gamma_pnl: float
    vega_pnl: float
    theta_pnl: float


class ScenarioRequest(BaseModel):
    account_id: str = "paper-account"
    underlying_shocks_pct: list[float] = Field(
        default=[-0.10, -0.05, 0, 0.05, 0.10], min_length=1, max_length=21
    )
    volatility_shocks_points: list[float] = Field(
        default=[-0.05, 0, 0.05], min_length=1, max_length=11
    )
    days_forward: list[int] = Field(default=[0, 1, 5], min_length=1, max_length=10)


class ExposureCell(BaseModel):
    symbol: str
    expiration: str
    delta: float
    gamma: float
    theta: float
    vega: float
    market_value: float
    unrealized_pnl: float


class RebalanceConstraint(BaseModel):
    max_abs_delta: float = Field(default=100, ge=0)
    max_abs_vega: float = Field(default=1000, ge=0)
    max_symbol_exposure_pct: float = Field(default=0.30, gt=0, le=1)
    account_equity: float = Field(default=100000, gt=0)


class RebalanceProposal(BaseModel):
    proposal_id: str
    created_at: datetime
    account_id: str
    status: Literal["PROPOSED"] = "PROPOSED"
    objective: str
    current_delta: float
    target_delta: float
    estimated_delta_change: float
    estimated_notional: float
    symbol: str
    side: Literal["BUY", "SELL"]
    quantity: int
    rationale: list[str]
    warnings: list[str]
    phase3_handoff: dict

File: src/trading_ai/ui/static/interactive_portfolio.js
(() => {
  const api=async(url,opt={})=>{const r=await fetch(url,{headers:{"Content-Type":"application/json"},cache:"no-store",...opt});const b=await r.json().catch(()=>({}));if(!r.ok)throw new Error(b.detail||JSON.stringify(b));return b};
  const n=(v,d=2)=>Number(v||0).toFixed(d);
  function nav(){
    const root=document.querySelector("#nav")||document.querySelector(".nav");
    if(root&&!root.querySelector('[data-view="interactive-portfolio"]')){
      const b=document.createElement("button");b.dataset.view="interactive-portfolio";b.textContent="Portfolio Manager";root.appendChild(b);
    }
  }
  async function render(){
    const title=document.querySelector("#pageTitle"),notice=document.querySelector("#notice"),content=document.querySelector("#content");
    if(!content)return;
    if(title)title.textContent="Interactive Portfolio Management";
    if(notice)notice.textContent="Analytics are read-only. Rebalancing creates a governed proposal for Phase 3 review and approval.";
    content.innerHTML="<p>Loading portfolio…</p>";
    try{
      const [s,m]=await Promise.all([api("/api/v1/interactive-portfolio/summary"),api("/api/v1/interactive-portfolio/exposure-matrix")]);
      content.innerHTML=`
      <div class="option-chain-summary">
        <div class="option-chain-card"><small>Market Value</small><strong>$${n(s.total_market_value)}</strong></div>
        <div class="option-chain-card"><small>Unrealized P/L</small><strong>$${n(s.total_unrealized_pnl)}</strong></div>
        <div class="option-chain-card"><small>Net Delta</small><strong>${n(s.net_delta)}</strong></div>
        <div class="option-chain-card"><small>Net Vega</small><strong>${n(s.net_vega)}</strong></div>
      </div>
      <div class="card"><h2>Positions</h2><div style="overflow:auto"><table><thead><tr><th>Symbol</th><th>Contract</th><th>Qty</th><th>Mark</th><th>Value</th><th>P/L</th><th>Δ</th><th>Γ</th><th>Θ</th><th>Vega</th></tr></thead><tbody>
      ${s.positions.map(p=>`<tr><td>${p.symbol}</td><td>${p.option_expiry||"Equity"} ${p.option_strike||""} ${p.option_type||""}</td><td>${p.quantity}</td><td>${n(p.mark_price)}</td><td>${n(p.market_value)}</td><td>${n(p.unrealized_pnl)}</td><td>${n(p.delta,3)}</td><td>${n(p.gamma,3)}</td><td>${n(p.theta,3)}</td><td>${n(p.vega,3)}</td></tr>`).join("")}
      </tbody></table></div></div>
      <div class="card"><h2>Exposure Heatmap</h2><div id="heatmap" style="display:grid;grid-template-columns:repeat(auto-fit,minmax(150px,1fr));gap:8px">
      ${m.map(x=>`<div class="option-chain-card" style="opacity:${Math.min(1,.3+Math.abs(x.delta)/Math.max(1,...m.map(y=>Math.abs(y.delta))))}"><strong>${x.symbol}</strong><br><small>${x.expiration}</small><br>Δ ${n(x.delta)} · Vega ${n(x.vega)}<br>P/L $${n(x.unrealized_pnl)}</div>`).join("")}</div></div>
      <div class="card"><h2>Scenario Analysis</h2><button id="runScenarios">Run Standard Grid</button><div id="scenarioResult"></div></div>
      <div class="card"><h2>Governed Rebalancing</h2><label>Maximum absolute delta <input id="maxDelta" type="number" value="100"></label><button id="proposeRebalance">Create Proposal</button><pre id="rebalanceResult"></pre></div>`;
      document.querySelector("#runScenarios").onclick=async()=>{
        const rows=await api("/api/v1/interactive-portfolio/scenarios",{method:"POST",body:JSON.stringify({account_id:"paper-account",underlying_shocks_pct:[-.1,-.05,0,.05,.1],volatility_shocks_points:[-.05,0,.05],days_forward:[0,1,5]})});
        const worst=[...rows].sort((a,b)=>a.estimated_pnl-b.estimated_pnl)[0],best=[...rows].sort((a,b)=>b.estimated_pnl-a.estimated_pnl)[0];
        document.querySelector("#scenarioResult").innerHTML=`<p>Worst: $${n(worst.estimated_pnl)} at ${(worst.underlying_shock_pct*100).toFixed(0)}% / vol ${(worst.volatility_shock_points*100).toFixed(0)} pts / ${worst.days_forward}d</p><p>Best: $${n(best.estimated_pnl)}</p>`;
      };
      document.querySelector("#proposeRebalance").onclick=async()=>{
        const max=Number(document.querySelector("#maxDelta").value);
        const p=await api(`/api/v1/interactive-portfolio/rebalance-proposal?account_id=paper-account`,{method:"POST",body:JSON.stringify({max_abs_delta:max,max_abs_vega:1000,max_symbol_exposure_pct:.3,account_equity:100000})});
        document.querySelector("#rebalanceResult").textContent=JSON.stringify(p,null,2);
        sessionStorage.setItem("portfolioRebalanceProposal",JSON.stringify(p.phase3_handoff));
      };
    }catch(e){content.innerHTML=`<div class="error">${e.message}</div>`}
  }
  window.addEventListener("DOMContentLoaded",()=>{nav();if(new URLSearchParams(location.search).get("view")==="interactive-portfolio")render()});
  document.addEventListener("click",e=>{if(e.target.closest('[data-view="interactive-portfolio"]')){history.pushState({},"","/?view=interactive-portfolio");render()}});
})();
